### DISCUSSION TITLE
Experimental Features

### DISCUSSION THREAD

**Discussion by advplyr** [advplyr/audiobookshelf#75] [Ideas]
This is where notes on the current experimental features will be.

## Ebook features are no longer experimental as of [v2.2.23](https://github.com/advplyr/audiobookshelf/releases/tag/v2.2.23)

In v2.2.23 the experimental features setting was removed since there are currently no experimental features.

---
### Old notes on ereader development

<details>
## eBook Reader - Stage 3

As of [v1.4.11](https://github.com/advplyr/audiobookshelf/releases/tag/v1.4.11), supported formats are: `epub`, `mobi`, `azw3`, `pdf`, `cbr`, `cbz`
There are 4 different readers:
- **epub reader**: flip through pages, 2-pages on screen [`epub`]
- **mobi reader**: no pages, scroll only (images not fully working) [`mobi`/`azw3`]
- **pdf reader**: flip through pages 1 at a time [`pdf`]
- **comic reader**: flip through pages of images [`cbr`/`cbz`]

All readers are still very primitive at the moment. Looking for suggestions on all of the above.

![image](https://user-images.githubusercontent.com/67830747/137606680-79d1fb0c-d599-46ff-a83f-0c71cfe40863.png)

## eBook Reader - Stage 2

eBook only folders are supported as of [v1.4.4](https://github.com/advplyr/audiobookshelf/releases/tag/v1.4.4). If you keep your ebooks and audiobooks separate, this means you can create a new library and scan your ebooks. Like so:

![image](https://user-images.githubusercontent.com/67830747/136713955-08df0e17-f8d8-4bce-aca9-7edb104ed6ae.png)

* ebook only books will show a book icon instead of a play icon, which will open the ereader.
    * or just click the blue book icon (this icon is just for helping identify books with an epub for testing)

## eBook Reader - Stage 1

- Currently only supports `ePub`.
- Will show small icon on book cards that have supported ePub.

![image](https://user-images.githubusercontent.com/67830747/135729347-f7fe6cd8-a585-4c10-abb8-068d90bc1f0a.png)

- and a Read icon

![image](https://user-images.githubusercontent.com/67830747/135729375-487ff364-7e94-43e3-802b-6e3918ceaa28.png)
</details>

**Comment by binarygeek119** [comment]
not sure if possible but could you be able to set he first chapter in reader then the audio would advance the page automatically.
i do understand it wouldnt be precise kinda do fuzzy logic maybe?

**Reply by advplyr** [reply]
The only way I could see this being possible is if the audio gets translated to text in the background and tries to match it with the current location in the ebook.
With the rapid improvement in speech-to-text technology I'm sure this is possible, and could be a fun project in the future.

**Reply by keaganhilliard** [reply]
https://github.com/r4victor/syncabook
I haven't looked into it extensively but EPUB 3.0 is supposed to have support for audio. Not sure how relevant this is, but seems like a good starting point.

**Reply by befuddled-dude** [reply]
Looks like they've done some of the work already here with **aeneas**:  https://www.readbeyond.it/aeneas/ 

This aeneas tool can take text and audio and create a sync-map between the two.  I guess you'd have to pull the unformatted text out of an epub, use that with aeneas and the correponding audio files, and then map the audio timestamps back to the epub offsets.  In theory, once it's done, you should be able to map every five seconds or so of audio the correct paragraph in an ebook.  

This would be a big project, no doubt!

**Reply by ghost** [reply]
Amazon WhisperSync (whispernet?) is a commercial application of this. If you own the audible audiobook and bought the book on Amazon, and you have a compatible Amazon tablet, it works. Basically, you open Kindle eBook and it will detect if you have the corresponding audiobook, if you do, you have the option to begin listening to the audiobook. Then at the same time the text that is currently being narrataed is highlighted, three or four words at a time. It's also called Immersive Reading. It's pretty awesome and I've been looking for an open source version for a long time. If this tool could somehow implement what I just described, that would be amazing.

**Reply by NLZ** [reply]
Storyteller acomplishes this as well

https://gitlab.com/smoores/storyteller

**Comment by binarygeek119** [comment]
Can we get a save for turning on experimental features and just have it on tell you turn it off.

Every time I reload I have to go in and turn it on.

**Reply by advplyr** [reply]
Experimental version toggle now persists in browser for [v1.4.1](https://github.com/advplyr/audiobookshelf/releases/tag/v1.4.1)

**Comment by Merijeek** [comment]
Silly question - if I turn on the button for the advanced features...should
I be looking for them in any particular location?

**Reply by advplyr** [reply]
Right now the only experimental feature is the ebook reader.
If you have an `epub` file in any of your audiobook directories you should see the blue icon like in the screenshot above.

In a coming release directories with only ebooks will be supported, for those that separate their audiobook and ebook libraries. At the moment saving and reloading backups is priority though.

**Comment by Merijeek** [comment]
Awesome, thanks.

I was going to suggest adding ebook functionality, but that seemed a bit
much to ask at this point.

But once you DO have it added, the ability to push it to an ereader ala
Calibre would be sooooo awesome.

**Reply by Jdiesel87** [reply]
Agreed this would be the major feature keeping me with Calibre. The OPDS server that allows me to send books directly to my Kobo reader is so convenient.

**Reply by advplyr** [reply]
Calibre is really the only game in town when it comes to converting between ebook formats and sending to devices. Right now I have calibre and audiobookshelf pointing to the same books folder, this way I have a nice view of my books, while calibre can handle the formatting and device syncing.
I'm not familiar with Kobo, but I know for kindle sending ebooks is as simple as sending an email.

**Reply by LostOnTheLine** [reply]
> I know for kindle sending ebooks is as simple as sending an email.

That's an option but definitely not the preferred method for many. Sending an eBook via EMail results in it being auto-converted & shrunk by Amazon & I've seen some books look bad afterwards. I don't think there will ever really be a convenient way to send to Kindle with Audiobookshelf other than that though as a USB cable from a web interface seems like an issue that can't really be solved.

**Reply by fivestones** [reply]
@advplyr not sure if you're still pointing calibre and audiobookshelf to the same books folder now 2 years later, but if so: How do you make this work? I've just tried the same thing, and while the books come in, Calibre adds its book id number to each book's directory name in parentheses, and when the books are scanned in audiobookshelf, I'm getting these id numbers (e.g., "book name (232)") as the book name in audiobookshelf.

I did turn on "prefer OPF metadata" in settings, but it doesn't seem to make a difference, even though each book folder has an OPF file with the correct metadata. Also, for some books genres are getting populated from the OPF file and for some they aren't--not sure why. What am I doing wrong? Maybe I just need to open a new issue.

**Reply by advplyr** [reply]
For my production server I don't use the calibre folder anymore, but my test server still does. Abs recognizes most of the metadata inside the OPF file so this should work correctly.
There are some downsides of the scanner that I hope to find a way to smooth out. If you did not have "Prefer OPF metadata" toggled when you did the first library scan then the OPF file is not going to override any metadata after that.
If you are okay with starting fresh then I suggest you delete the library from Abs. Make sure you have "Prefer OPF metadata" toggled on. Then add a new library and scan.
I tested this just now to make sure it works properly.

**Reply by fivestones** [reply]
Awesome. That did the trick. They all nicely imported now.
It would be nice to be able to do some kind of scan that looks for updated opf metadata and uses that for the book's metadata. Not sure if this is something on the radar for you. I noticed at least one book in calibre didn't have any metadata, and so there was no opf file. After importing it into audiobookshelf, I updated the metadata in calibre, and then went back to audiobookshelf and edited the book and did scan again, and it pulled in the new metadata from the opf file.
So maybe if you start with no opf file, then add one, you can rescan and get this info, but if you started with an opf file and then update it, you won't be able to pull in new information after updating it? I'm not sure how this works. But if you have the setting to prefer opf metadata, it seems like a scan should look at any current existing opf file and use that metadata.
Just some thoughts.
Thanks for your help! And for this awesome software!

**Reply by advplyr** [reply]
This issue was the tipping point for a big refactor of the scanner. There is a lot that goes into the scanning process so I can't break it all down here but the next release will include documentation explaining everything. I was just posting about it in Discord

**Comment by feerlessleadr** [comment]
I'm not sure if this is a bug, or the feature is not implemented yet, but I'm trying this out now, and there are 2 items that I think are needed to fit my use case (one may be a bug, the other is definitely a feature request):

- Save reading position of ebook files (in my testing, my reading position for the ebook files is not saved, and so every time I close and re-open a book, it starts from the beginning)
- Ability to change the font  size of the ebook text in the reader (make the text bigger or smaller)

Thanks!

**Reply by vincentscode** [reply]
Being able to resume reading where you left off would be amazing. Without it it get's rather annoying really fast.

**Reply by advplyr** [reply]
Correct, I wouldn't recommend this as an e-reader at this point.

**Comment by AlexandreBarde** [comment]
Hi ! 
I can't read an ebook (.epub) on the IOS app. When I click on "read", the cover opens but I can't navigate through the pages.
I don't have any buttons, like "<", ">" to switch between pages.

Maybe, I don't using it correclty ?

Thanks ;)

**Reply by MaestroJAL** [reply]
The reader works better in iOS Safari. I can see pages and turn them with the arrows but the original cover art is superimposed over the pages making reading difficult. I can move the cover art layer up and down to some extent, but can't hide it to get to the content under it. The app seems to stretch beyond the viewing capacity of the iPhone screen. Maybe this helps?

**Reply by advplyr** [reply]
@MaestroJAL Can you share a screenshot of what you mean?

**Reply by MaestroJAL** [reply]
Sure! Here's the App version for reading an ePub:
![image](https://user-images.githubusercontent.com/12278867/221047894-b0f3e830-80c6-4ad3-a829-18f607d31b22.png)

Here's the Safari version for the same book:
![image](https://user-images.githubusercontent.com/12278867/221048461-298ec347-b5f6-453e-b55a-83cd0dc64f66.jpeg)

**Reply by advplyr** [reply]
What version of the server are you running?

**Reply by MaestroJAL** [reply]
I'm using ABS 2.2.15 via Docker. I also have 0.9.62-beta of the App.

**Reply by MaestroJAL** [reply]
Progress and usability! Thank you! The Safari web app still superimposes the cover over the reading layer but it now can be moved completely out of the way allowing the user to navigate the book. The iOS app works fairly well! Swiping the screen to navigate from page to page works well! MINOR issue: On my iPhone 12 Pro Max, the progress is slightly hidden by the restraints of the rounded bezels. Frankly, it doesn't really matter. I can see the progress. I'm using ABS Server 2.2.16 via Docker and version 0.9.63 of the iOS app.

Safari Web App:
![image](https://user-images.githubusercontent.com/12278867/223570269-2b0536e5-dfbe-4fef-b949-18f762abf5cc.jpeg) 

ABS iOS App:
![image](https://user-images.githubusercontent.com/12278867/223570412-f0926795-8e19-489a-9a7a-0ec13d6066c4.png)

**Comment by Telkhine** [comment]
A way to fullscreen the reader would be nice. There is currently a lot of blank space around the book. It's especially bad on my 2k Monitor:
![Screenshot (52)](https://user-images.githubusercontent.com/78505/173188906-df417848-449f-4e62-b327-0249dfb32051.png)
![Screenshot (53)](https://user-images.githubusercontent.com/78505/173188910-35ac06c5-974b-4885-91f3-0d59b349a9cd.png)

**Comment by Ubersjaak** [comment]
- It'd be of more use if it actually remembered your progress
- Adding bookmarks would be awesome

**Comment by ghost** [comment]
I would love to see the full corpus of the epub library, indexed and searchable. For example: http://protected-dusk-3051.herokuapp.com/
A demo based on this project: https://github.com/larsvoigt/epub-full-text-search
I'm seeing some other project provide search to epub libraries, please see: 
https://github.com/simon987/sist2
https://github.com/e-alfred/epubreader - also allows for highlighting, bookmarking, auto-tag keywords. looks like it might index comic CBZs as well
Also, would allowing conversion between different file formats be possible?
Thank you for everyone's hard work. This is an amazing community and Audiobookshelf brings me much joy everyday.

**Comment by LostOnTheLine** [comment]
A simple addition, but one that shouldn't be complicated, can you add an option for "Manga Mode" like most readers have? for my case it's CBZs, I know ePubs have a similar option that can be encoded into the ePub but some readers will ignore them, don't have any to test so I don't know if this one does or not.

Essentially all it is is a page turn direction flip.
For manga, where the page read direction is upper-right to lower left flipping pages the other way can be annoying.

I'm loving the app by the way. & the support for CBZ is great because your interface is better than the current comic/manga servers I've tried & their organization requirements are always very specific & never work from one to the other so with the page flip this will cover a lot & it will make the things for the kids all in one place.

**Reply by advplyr** [reply]
So this doesn't get lost here I created a feature request for this https://github.com/advplyr/audiobookshelf/issues/1274

**Comment by hellofaduck** [comment]
Any plans to add fb2 e-book format to reader? It's most popular ebook format over all CIS countries, of course converters from fb2 to other format exists but it's always a pleasure to have native experience without converters shenanigans.

**Comment by scottdk** [comment]
Where are the ebook files stored locally on Android? 

I added a mobi file to the audiobook folder on the server, but nothing comes up when I search my phone's file system for "mobi" or the book's name. Also nothing under com.audiobookshelf.app on  the Android/media and /data folders.

Android app version 0.9.63-beta.

**Comment by maulik-sharma** [comment]
I would really love a send to kindle option and an option to convert e-books into different formats like you can in calibre.

**Comment by nickkimbrough** [comment]
Something that basically all of the popular self hosted e-readers apps are missing is the ability to read along with audio on EPUB3 files. Some good examples of software that DO do this are [Thorium](https://www.edrlab.org/software/thorium-reader/) and [ReadBeyond](https://www.readbeyond.it/ebooks.html) (no longer supported). The paid closed source alternative being Amazon WhisperSync.

There are no popular self hosted options that do this to speak of, this is a big open spot in the market, whereas apps that read epub and other files are pretty common. I also think it fits in nicely with Audio Bookshelf being primarily audio focused. The primary reason I picked up Audiobookshelf was due to nobody having this ability, otherwise I would just load up the EPUB3 and listen and read along if I wanted.

I think this is called 'media overlay' and specified in the EPUB3.3 definition: https://www.w3.org/TR/epub-33/#sec-media-overlays

**Reply by advplyr** [reply]
This has been mentioned in #189 and #1723

We've also talked about this in Discord a few times. It would be a really cool feature. Right now this seems like a very big undertaking so I probably won't be building this anytime soon. Audible has an advantage in that they don't have to deal with user-provided ebooks and audio files.
The epub spec could be helpful since it would be the epub that is supplying the synced timestamps and locations.
The most likely chance this gets added is if someone builds a node package that takes an ebook and audio files and returns locations/timestamps.

**Reply by ndevries84** [reply]
Someone just recently did this https://smoores.gitlab.io/storyteller/docs/what-is-this/  I create Epub3 with media overlays automatically.  You just need to select the audiobook and epub files and it will merge them into an epub3 with media overlay.

**Reply by fivestones** [reply]
> Someone just recently did this

...and it's MIT licensed!

**Reply by applesoff** [reply]
Now it's been almost a year. I am not on discord to know if additional thought has gone into this. Is something like this more doable now?

**Reply by nichwall** [reply]
There has not been progress on this in ABS.

**Reply by ambroserb3** [reply]
This would be a huge feature for audiobookshelf, is anyone working on it?

**Reply by tobiasredel** [reply]
maybe this does help someone: https://github.com/tobiasredel/shelfsync
